rpc_state_reader_sn_api/src/lib.rs

```diff
@@ -0,0 +1,994 @@
+
+impl<'de> DeserializeAs<'de, StarkFelt> for StarkFeltHex {
+    fn deserialize_as<D>(deserializer: D) -> Result<StarkFelt, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let bytes: PrefixedBytesAsHex<32> = PrefixedBytesAsHex::deserialize(deserializer)?;
+        Ok(bytes.try_into().unwrap())
+    }
+}
+
+#[serde_as]
+#[derive(Debug, Deserialize)]
+struct RpcResponseFelt {
+    #[serde_as(as = "StarkFeltHex")]
+    result: StarkFelt,
+}
+
+pub struct RpcBlockInfo {
+    /// The sequence number of the last block created.
+    pub block_number: BlockNumber,
+    /// Timestamp of the beginning of the last block creation attempt.
+    pub block_timestamp: BlockTimestamp,
+    /// The sequencer address of this block.
+    pub sequencer_address: ContractAddress,
+}
+
+type RpcResponse = ureq::Response;
+
+impl RpcState {
+    pub fn new(chain: RpcChain, block: BlockValue) -> Self {
+        if env::var("INFURA_API_KEY").is_err() {
+            dotenv().expect("Missing .env file");
+        }
+        Self {
+            chain,
+            api_key: env::var("INFURA_API_KEY")
+                .expect("Missing API Key in environment: INFURA_API_KEY"),
+            block,
+        }
+    }
+
+    fn rpc_call<T: for<'a> Deserialize<'a>>(
+        &self,
+        params: &serde_json::Value,
+    ) -> Result<T, RpcError> {
+        Self::deserialize_call(self.rpc_call_no_deserialize(params)?)
+    }
+
+    fn rpc_call_no_deserialize(&self, params: &serde_json::Value) -> Result<RpcResponse, RpcError> {
+        ureq::post(&format!(
+            "https://{}.infura.io/v3/{}",
+            self.chain, self.api_key
+        ))
+        .set("Content-Type", "application/json")
+        .set("accept", "application/json")
+        .send_json(params)
+        .map_err(|err| RpcError::Request(err.to_string()))
+    }
+
+    fn deserialize_call<T: for<'a> Deserialize<'a>>(response: RpcResponse) -> Result<T, RpcError> {
+        let response = response.into_string().map_err(|err| {
+            RpcError::Cast("Response".to_owned(), "String".to_owned(), err.to_string())
+        })?;
+        serde_json::from_str(&response).map_err(|err| RpcError::RpcCall(err.to_string()))
+    }
+}
+
+#[derive(Debug, Deserialize)]
+pub struct RpcResult<T> {
+    result: T,
+}
+
+#[derive(Debug)]
+pub struct TransactionTrace {
+    pub validate_invocation: RpcCallInfo,
+    pub function_invocation: RpcCallInfo,
+    pub fee_transfer_invocation: RpcCallInfo,
+    pub signature: Vec<StarkFelt>,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq)]
+pub struct RpcExecutionResources {
+    pub n_steps: usize,
+    pub n_memory_holes: usize,
+    pub builtin_instance_counter: HashMap<String, usize>,
+}
+
+#[derive(Debug)]
+pub struct RpcCallInfo {
+    pub execution_resources: VmExecutionResources,
+    pub retdata: Option<Vec<StarkFelt>>,
+    pub calldata: Option<Vec<StarkFelt>>,
+    pub internal_calls: Vec<RpcCallInfo>,
+}
+
+#[allow(unused)]
```

Comment:
I don't think this `#[allow(unused)]` is necessary anymore since the struct is externally public.

Reply:
The fields aren't, but there's no reason to privatize them. Originally I made everything public so we could use these types in the other RPC StateReader if necessary but now we decided to deprecate that one. I think for this PR the visibility can be left unchanged but in a next PR they may change.